bundles/org.openhab.binding.bluetooth.roaming/README.md

```diff
@@ -0,0 +1,41 @@
+# Bluetooth Roaming Adapter
+
+This extension adds support for accessing Bluetooth devices from any other configured adapter via a virtual adapter.
+
+## Supported Things
+
+It defines the following bridge type:
+
+| Bridge Type ID | Description                                                                                                |
+|----------------|------------------------------------------------------------------------------------------------------------|
+| roaming        | A virtual Bluetooth adapter that interacts with Bluetooth devices through their nearest Bluetooth adapter. |
+
+## Channels
+
+Devices which use a `roaming` adapter as their bridge also gain the following channels:
+
+| Channel ID       | Item Type | Description                                                     |
+|------------------|-----------|-----------------------------------------------------------------|
+| adapter-uid      | String    | The thingUID of the adapter that is nearest to this device      |
+| adapter-location | String    | The value of the `Location` specified for the nearest adapter    |
+
+## Discovery
+
+A default roaming adapter is discovered automatically, although you can still configure it textually if you want.
+There can be only a single roaming adapter on a system.
+
+## Bridge Configuration
+
+The roaming bridge requires an `address` parameter which mearly servers as an identifier for the virtual adapter, its value doesn't really matter. 
+It is advised to make sure it is set to a value outside of the normal bluetooth mac address range, for example: "FF:FF:FF:FF:FF:FF".
```

Comment:
This sounds as if the value should be a mac address. Could it also be `/(%§&@~$")` or is its syntax checked against mac structure?

Reply:
Right now the value is required to be a mac address to create an instance of `BluetoothAddress` for the adapter's `getAddress()` method to return.

Funnily enough, after looking through the Bluetooth binding code it seems that only place that BluetoothAdapter.getAddress() is even called is in a logging statement. So there isn't really any architectural requirement for `BluetoothAdapter` to even have a getAddress method.
Would it make sense to remove that method from the `BluetoothAdapter` interface to allow users to create roaming thing instances without having to specify a fake bluetooth mac address?

Reply:
Well, this parameter is what makes the bridge unique - it is the only differentiator between two bridges. Just as you say "it merely serves as an identifier". But this means we cannot remove it (or if we did, we would need to add some other parameter as a replacement). So I guess we can keep it and should only specify, what the value of the address should be for roaming adapters. If it does not have to be a mac, you could come up with something yourself, like saying it has to be `virtual:XX` with XX a byte value or whatever. And having a small syntax check on the parameter then probably won't hurt.

Reply:
> If it does not have to be a mac, you could come up with something yourself, like saying it has to be virtual:XX with XX a byte value or whatever. And having a small syntax check on the parameter then probably won't hurt.

Well right now it has to be a mac in order to be able to implement the `BluetoothAdapter.getAddress()` method but if that method were to be removed you are correct that the identifier could be anything we want.
But it might also be possible to not have any identifier for roaming bridges at all outside of their thingUID. Roaming bridges only exist as ThingHandlers and have no actual backing device; they exist as soon as a roaming thing requests a handler.

Is the ThingUID not a sufficient bridge identifier? Or do you think a separate identifier is required?

Reply:
> right now it has to be a mac in order to be able to implement the BluetoothAdapter.getAddress()

But what you are saying is that this method isn't really used anywhere. So why does it have to be a mac when it just says "address"? My idea was to define "address" a bit wider than being a mac and thus use something artificial like `virtual:01` as an address for the roaming bridge.

> Is the ThingUID not a sufficient bridge identifier?

No, as mentioned above, it would mean that it is the very same "thing", just being referenced with different ids. For the roaming bridge it is difficult to argue, since it is an exceptional abstract case. I somehow feel that we nonetheless should not make any exceptions for it, at least when it isn't a big deal for us.

Reply:
> So why does it have to be a mac when it just says "address"?

It is meant to be reflective of the how the bluetooth protocols are meant to operate. All devices on a bluetooth network have a mac address, host and devices, through which they identify themselves as they communicate across the network. The `BluetoothAdapter.getAddress` method was meant to reflect an attribute of real physical bluetooth adapters which have unique mac addresses.

All `BluetoothAdapter` implementations are required to return a non-null `BluetoothAddress` object from their `BluetoothAdapter.getAddress` implementation.

Because the roaming adapter has to implement `BluetoothAdapter` then it must also be able to provide an BluetoothAddress instance to return. The `BluetoothAddress` enforces the mac format of the address on creation, so their is no way to not have a BluetoothAddress that isn't in the XX:XX:XX:XX:XX:XX format.

So if we want the roaming adapter to support non-mac addresses then we would either have to remove `BluetoothAdapter.getAddress`, change `BluetoothAdapter.getAddress`'s return type to be nullable so that the roaming adapter can return null, change the return type to something that isn't `BluetoothAddress`, or modify `BluetoothAddress` to allow non-mac address formats.
Which one do you think I should go with?

Reply:
I just read that devices are often using random MAC addresses for privacy reasons. It therefore seems to me that it is legit to just provide ANY MAC address - so your suggestion in the readme to just define any MAC address as long as it is in the valid format might indeed be the simplest way to go?
You say "a value outside of the normal bluetooth mac address range" - can this be specified in more detail?

Reply:
@cpmeister If you want to get it into 2.5.7, we should now finish it soon.
Would you agree with my comment above? If so, please adapt the code, rebase it and we should be good to merge!

Reply:
> You say "a value outside of the normal bluetooth mac address range" - can this be specified in more detail?

You give one example now ("FF:FF:FF:FF:FF:FF"), but "value outside normal range" sounds like there should be a kind of "reserved" address space for local/private addresses. What values other than "FF:FF:FF:FF:FF:FF" would you think are ok to use?

Reply:
I went ahead and just removed the address from the roaming adapters entirely. It wasn't meaningfully used in the code and would only confuse users trying to configure a roaming adapter thing.

Reply:
Hm, it seems you forgot about all previous discussion we had here - removing the address was not really an option, see https://github.com/openhab/openhab-addons/pull/7199#discussion_r439770589.

Reply:
> Hm, it seems you forgot about all previous discussion we had here - removing the address was not really an option, see #7199 (comment).

Apparently I did forget! But since I loathe idea of a useless configuration parameter I went ahead and implemented the roaming adapter groups feature. This feature adds a `groupUIDs` parameter which could serve as a thing identifier for any roaming bridge thing. I've also eliminated the restriction of only allowing a single roaming adapter thing to be defined.

I hope this is sufficient compromise.

Reply:
Ok, sounds like a decent compromise, thank you!